library/std/src/sys/pal/windows/c/windows_sys.rs

```diff
@@ -3847,21 +3845,21 @@ pub type SET_FILE_POINTER_MOVE_METHOD = u32;
 #[repr(C)]
 pub struct SOCKADDR {
     pub sa_family: ADDRESS_FAMILY,
-    pub sa_data: [u8; 14],
+    pub sa_data: [i8; 14],
 }
-impl ::core::marker::Copy for SOCKADDR {}
-impl ::core::clone::Clone for SOCKADDR {
+impl Copy for SOCKADDR {}
+impl Clone for SOCKADDR {
     fn clone(&self) -> Self {
         *self
     }
 }
 #[repr(C)]
 pub struct SOCKADDR_UN {
     pub sun_family: ADDRESS_FAMILY,
-    pub sun_path: [u8; 108],
+    pub sun_path: [i8; 108],
```

Comment:
I believe this is also supposed to be `ffi::c_char`? Is there a reason that windows-bindgen is lowering these to  `i8` instead?

Reply:
Well I'd guess it's because `ffi::c_char` is defined as `i8` on Windows 😉. It's just that Rust tends to assume bytes are `u8`.

Reply:
https://github.com/rust-lang/rust/blob/6b1e5d9db3b0d569bd847fdcb98c85ee8f7a8ef1/library/core/src/ffi/mod.rs#L83

Reply:
I don't know if that's correct for AArch64.

Reply:
On AArch64 MSVC C++ `std::is_signed_v<char>` is `true` and `CHAR_MIN` is `-128`.

Well there is a [C/C++ compiler option](https://learn.microsoft.com/en-us/cpp/build/reference/j-default-char-type-is-unsigned?view=msvc-170) to make it signed but that's not set by default and in any case Rust has no similar option to choose the signedness so going with the default makes sense.

Reply:
Huh.

Reply:
Ah, there is some documentation. You can infer from the absence of target-specific information that it's the same across targets: https://learn.microsoft.com/en-us/cpp/c-language/data-type-specifiers-and-equivalents?view=msvc-170

Reply:
@ChrisDenton I thought AArch64 Windows was supposed to use the AAPCS64 which specifies that `char` is `unsigned`:
- https://learn.microsoft.com/en-us/cpp/build/arm64-windows-abi-conventions?view=msvc-170

https://github.com/ARM-software/abi-aa/blob/879df44cb00de276a74034180adf61996b03afda/aapcs64/aapcs64.rst#L2441-L2449

Reply:
These documents conflict.

Reply:
Well, not necessarily...

> A platform ABI may specify a different combination of primitive variants but we discourage this.

And i can know for sure what Visual Studio does when you compile an application for ARM64

Reply:
It's not like it's the first time MSVC did something weird in the name of compatibility with its older stuff.

Reply:
Of course, it is not entirely impermissible. But the article I linked specifically says that it documents the variations from the AArch64 EABI, and it doesn't mention anything about `char`, which means that whatever this is, it sure as hell ain't **specified**.

Reply:
It wouldn't even be the first undocumented deviation from AAPCS! That doc also says

> For non-variadic functions, the Windows ABI follows the rules specified by ARM for parameter passing.

which I [happen](https://github.com/rust-lang/rust/pull/112157#issuecomment-1588474831) to know is...a creative representation of the truth.

<sup>Ironically, Microsoft's calling convention has the least special cases, missing a few from Darwin, and missing [this](https://github.com/rust-lang/rust/blob/a165f1f65015b1bd4afd2ec50700aaacf2e0c485/compiler/rustc_target/src/abi/call/aarch64.rs#L92-L98) incredibly cursed [special case](https://github.com/ARM-software/abi-aa/blob/879df44cb00de276a74034180adf61996b03afda/aapcs64/aapcs64.rst#L1860-L1871) from AAPCS that ignores requested alignment in some situations.</sup>